File: moviereview/src/main/java/com/abc/moviereview/service/ReviewService.java
package com.abc.moviereview.service;

import com.abc.moviereview.dto.ReviewDTO;

import java.util.List;
import java.util.Optional;

/**
 * Service interface for Review operations.
 * Provides business logic for CRUD operations on reviews.
 */
public interface ReviewService {

    /**
     * Get all reviews.
     * @return List of all reviews
     */
    List<ReviewDTO> getAllReviews();

    /**
     * Get a review by its ID.
     * @param id Review ID
     * @return Optional containing the review if found, empty otherwise
     */
    Optional<ReviewDTO> getReviewById(Long id);

    /**
     * Get all reviews for a movie.
     * @param movieId Movie ID
     * @return List of reviews for the movie
     */
    List<ReviewDTO> getReviewsByMovieId(Long movieId);

    /**
     * Create a new review.
     * @param reviewDTO Review to create
     * @return Created review
     */
    ReviewDTO createReview(ReviewDTO reviewDTO);

    /**
     * Update an existing review.
     * @param id Review ID
     * @param reviewDTO Updated review details
     * @return Updated review if found, null otherwise
     */
    ReviewDTO updateReview(Long id, ReviewDTO reviewDTO);

    /**
     * Delete a review.
     * @param id Review ID
     * @return true if deleted, false if not found
     */
    boolean deleteReview(Long id);
}

File: movieapp/src/app/services/review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Review } from '../models/review.model';
import { environment } from '../../environments/environment';
import { NGXLogger } from 'ngx-logger';

@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  private apiUrl = environment.movieReviewApiUrl;

  constructor(
    private http: HttpClient,
    private logger: NGXLogger
  ) {}

  /**
   * Get all reviews
   * @returns Observable<Review[]>
   */
  getAllReviews(): Observable<Review[]> {
    this.logger.debug('Fetching all reviews');
    return this.http.get<Review[]>(this.apiUrl);
  }

  /**
   * Get review by ID
   * @param id Review ID
   * @returns Observable<Review>
   */
  getReviewById(id: number): Observable<Review> {
    this.logger.debug(`Fetching review with ID: ${id}`);
    return this.http.get<Review>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get reviews by movie ID
   * @param movieId Movie ID
   * @returns Observable<Review[]>
   */
  getReviewsByMovieId(movieId: number): Observable<Review[]> {
    this.logger.debug(`Fetching reviews for movie with ID: ${movieId}`);
    return this.http.get<Review[]>(`${this.apiUrl}/movie/${movieId}`);
  }

  /**
   * Create a new review
   * @param review Review data
   * @returns Observable<Review>
   */
  createReview(review: Review): Observable<Review> {
    this.logger.debug('Creating new review', review);
    return this.http.post<Review>(this.apiUrl, review);
  }

  /**
   * Update an existing review
   * @param id Review ID
   * @param review Updated review data
   * @returns Observable<Review>
   */
  updateReview(id: number, review: Review): Observable<Review> {
    this.logger.debug(`Updating review with ID: ${id}`, review);
    return this.http.put<Review>(`${this.apiUrl}/${id}`, review);
  }

  /**
   * Delete a review
   * @param id Review ID
   * @returns Observable<void>
   */
  deleteReview(id: number): Observable<void> {
    this.logger.debug(`Deleting review with ID: ${id}`);
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
